#96 Enable specifying auth at the Service definition level

chankeypathak opened the pull request:
**This change is a:** (check at least one)
- [ ] Bugfix
- [x] Feature addition
- [ ] Code style update
- [ ] Refactor
- [ ] Release activity

**Is this a breaking change?** (check one)
- [ ] Yes
- [x] No

**Is the:**
- [x] Title of this pull request clear, concise, and indicative of the issue number it addresses, if any?
- [x] Test suite passing?
- [x] Code coverage maximal?
- [x] Changelog up to date?

**What does this change address?**
Resolves #96 

**How does this change work?**
By passing auth argument to underlying requests' methods

chankeypathak committed:
#96 add auth in Service

chankeypathak committed:
#96 fix black

daneah committed:
Merge remote-tracking branch 'upstream/dev' into dev

* upstream/dev:

daneah committed:
Fix tests and code formatting

The tox lint environment wasn't set up to _act_ on linting errors, so I
updated the commands. They can be run with the same command as prior for
the same behavior as prior, but can also be run with a path now to
actually specify the directory or format the code:

```shell
$ tox -e lint src tests
```

on-behalf-of: @ithaka <dane.hillard@ithaka.org>

daneah reviewed (approved):
This looks good to go. I tried to add some documentation improvements as well because I realized that the [service docs](https://apiron.readthedocs.io/en/latest/api-reference/services.html) don't make mention of these class attributes we've been adding, but I believe I ran into [this bug](https://github.com/sphinx-doc/sphinx/issues/741) along the way and couldn't figure out a great way around it. I could've alternatively documented `ServiceBase` so that it could show those, but I wanted to avoid exposing what's supposed to be an internally-used object.

daneah committed:
Update changelog

on-behalf-of: @ithaka <dane.hillard@ithaka.org>

daneah committed:
Allow passed-in auth to take precedence

Previous changes blocked existing functionality that allowed an `auth`
kwarg to be passed into `call`. This fixes that, so the precedence is
now:

1. Passed-in auth kwarg
1. Passed-in session auth
1. Service auth

This makes the most explicitly-defined auth take highest priority, with
the auth defined furthest from the call taking lowest priority.

on-behalf-of: @ithaka <dane.hillard@ithaka.org>

daneah commented:
I pushed one more notable change. The change as submitted blocked existing functionality that allowed an `auth`
kwarg to be passed into `call`. This fixed that, so the precedence is now:

1. Passed-in `auth` kwarg
1. Passed-in `session`'s `auth`
1. Service's `auth`

This makes the most explicitly-defined `auth` take highest priority, with the `auth` defined furthest from the call taking lowest priority.

I may see if I can extricate the tests so that we can write tests for these cases more easily—the fact that I missed this certainly means the code is getting complex enough, and the test coverage alone isn't sufficient 😄

daneah committed:
Add tests for call auth priority

* Install pytest-randomly to uncover test ordering issues
* Remove `-q` flag when calling pytest to make test setup summary
  visible
* Remove class scope from test fixture and remove class wrapper around
  tests to ensure their independence

on-behalf-of: @ithaka <dane.hillard@ithaka.org>

daneah committed:
Remove unused import

on-behalf-of: @ithaka <dane.hillard@ithaka.org>

daneah commented:
The diff got rather shredded up on the test module, but [here's the test case](https://github.com/ithaka/apiron/blob/4dde43a1f39215bd7f1560279551db7fcb69b83f/tests/test_client.py#L181-L206) I added.

daneah committed:
Clean up logging mocks in tests

on-behalf-of: @ithaka <dane.hillard@ithaka.org>

daneah committed:
Merge remote-tracking branch 'upstream/dev' into dev

* upstream/dev:
  Add docstring for new parameter
  Enable raw response object in Endpoint definition

NazimHAli commented on setup.cfg:
```diff
@@ -44,6 +44,7 @@ lint =
 test =
     pytest==5.2.2
     pytest-cov==2.8.1
+    pytest-randomly==3.4.0
```
🎉

NazimHAli reviewed (approved):
Thank you for the contribution, looks good! 🎉
